pkg/exchange/okex/parse.go

```diff
@@ -8,59 +8,66 @@ import (
 	"strings"
 	"time"
 
-	"github.com/valyala/fastjson"
-
 	"github.com/c9s/bbgo/pkg/exchange/okex/okexapi"
 	"github.com/c9s/bbgo/pkg/fixedpoint"
 	"github.com/c9s/bbgo/pkg/types"
 )
 
-func parseWebSocketEvent(str []byte) (interface{}, error) {
-	v, err := fastjson.ParseBytes(str)
+func parseWebSocketEvent(in []byte) (interface{}, error) {
+	var e WsEvent
+
+	err := json.Unmarshal(in, &e)
 	if err != nil {
 		return nil, err
 	}
 
-	if v.Exists("event") {
-		return parseEvent(v)
-	}
+	switch {
+	case e.IsOp():
+		return e.WebSocketOpEvent, nil
+	case e.IsPushDataEvent():
+		// need unmarshal again because arg in both WebSocketOpEvent and WebSocketPushDataEvent
+		var pushDataEvent WebSocketPushDataEvent
 
-	if v.Exists("data") {
-		return parseData(v)
-	}
+		err := json.Unmarshal(in, &pushDataEvent)
+		if err != nil {
+			return nil, err
+		}
 
-	return nil, nil
-}
+		channel := pushDataEvent.Arg.Channel
 
-type WebSocketEvent struct {
-	Event   string      `json:"event"`
-	Code    string      `json:"code,omitempty"`
-	Message string      `json:"msg,omitempty"`
-	Arg     interface{} `json:"arg,omitempty"`
-}
+		switch channel {
 
-func parseEvent(v *fastjson.Value) (*WebSocketEvent, error) {
-	// event could be "subscribe", "unsubscribe" or "error"
-	event := string(v.GetStringBytes("event"))
-	code := string(v.GetStringBytes("code"))
-	message := string(v.GetStringBytes("msg"))
-	arg := v.GetObject("arg")
-	return &WebSocketEvent{
-		Event:   event,
-		Code:    code,
-		Message: message,
-		Arg:     arg,
-	}, nil
+		case string(WsChannelTypeBooks5):
+			data, err := parseBookData(&pushDataEvent)
```

Comment:
you can return ` parseBookData(&pushDataEvent)` directly

Reply:
will update